include/deal.II/lac/la_parallel_block_vector.h

```diff
@@ -348,10 +348,24 @@ namespace LinearAlgebra
        * After calling this method, read access to ghost elements of the
        * vector is forbidden and an exception is thrown. Only write access to
        * ghost elements is allowed in this state.
+       *
+       * @deprecated Use zero_out_ghost_values() instead.
        */
       void
       zero_out_ghosts() const;
```

Comment:
Can you mark up this function with `DEAL_II_DEPRECATED`?

Reply:
Ah, I'm only now reading @kronbichler 's comment.

How do we track these things? If we don't `DEAL_II_DEPRECATE` these functions right away, then we should at least have an issue somewhere that collects all of those symbols we'd like to deprecate after the next release...

Reply:
Yes, also `contrib/utilities/print_deprecated.py` always searches for DEAL_II_DEPRECATED.
I don't think we really came to a conclusion regarding our policy to not deprecate immediately...
I still don't see a problem if codes that use `upstream` just always disable deprecation warnings if they don't want to deal with them.

Reply:
I am probably biased on this topic because I see the cost of taking care of these warnings in maybe half a dozen projects and I work with upstream. But it is not just upstream (how large is the percentage of users on master vs  a release, by the way?). I think many projects gravitate around two subsequent releases of deal.II. For me it is the latest release and master, but it does not really change much from a user perspective if you work with say the 9.2 and 9.1 releases. If you do not want warnings that came up in 9.2 and where 9.1 did not yet provide an alternative, you need all users of a project using deal.II to update at the same time, unless you ifdef by version. For small functions, ifdef is fine, but for more widely used interfaces it is a pain. So to me the important part is that we have one release with the new alternative and without the warning. Note that I am not opposed to eventually deprecate, but a distance of two releases is much easier to handle, as I see no problem to force users not being more than one release behind whatever project XY sees as the most recent deal.II version.

To me, the question is what we think makes best value out of the deprecation warnings. If we are too aggressive, we force people to either #ifdef by version, or to disable the warnings in all their builds and only be told later by an error (in which case we do not need the warnings). I think both alternatives have a good potential to frustrate users. If we are too soft, we have a maintainability problem, and this is not what I want. That's why I am in favor of an intermediate level for *central functionality* that has the potential to raise dozens of warnings (and make users disable the warning, rather than fix it). Of course, what is *central* is subjective and again puts a maintainability burden on us. I see me doing less work on keeping things up in deal.II than in the user projects I have on deal.II, but I realize that you @masterleinad have done much more cleaning work than me, so I admit that I'm not in a neutral position on this point.

As it seems I am the only person arguing this way, I assume that it is my own work load and not a general problem, so I better keep silent on this now.

Reply:
@kronbichler I see your point and agree with you concerns (I might have been too quick in deprecating since I got the impression that that is the normal: `DoFHandler::initialize()` , `VectorizedArray::n_vector_entries`). I think we were considering a version number of the `DEAL_II_DEPRECATED` macro. Maybe it is a good time to refresh the discussion there. I have a (maybe naive) implementation which might work (and could be configured at compile time via `DEAL_II_DEPRECATED_VERBOSITY`):

```cpp
#define DEAL_II_DEPRECATED_VERBOSITY 0
#define DEAL_II_DEPRECATED "[[deprecated]]"

#define CAT(a, ...) PRIMITIVE_CAT(a, __VA_ARGS__)
#define PRIMITIVE_CAT(a, ...) a ## __VA_ARGS__
#define DEAL_II_DEPRECATED_VERSION(major,minor) PRIMITIVE_CAT(DEAL_II_DEPRECATED_VERSION_, major ## _ ## minor)

#if DEAL_II_DEPRECATED_VERBOSE
  #define DEAL_II_DEPRECATED_VERSION_9_3 DEAL_II_DEPRECATED
#else
  #define DEAL_II_DEPRECATED_VERSION_9_3 ""
#endif

#define DEAL_II_DEPRECATED_VERSION_9_2 DEAL_II_DEPRECATED

#include <iostream>

int main()
{
    std::cout << DEAL_II_DEPRECATED_VERSION(9,2) << std::endl; // [[deprecated]]
    std::cout << DEAL_II_DEPRECATED_VERSION(9,3) << std::endl; //
}
```

... this was the best I could come up with the help of https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms. I would have loved if I didn't need to write out all version but could use the logic of:
```
#  define DEAL_II_MPI_VERSION_GTE(major,minor) \
 ((DEAL_II_MPI_VERSION_MAJOR * 100 + DEAL_II_MPI_VERSION_MINOR) >= (major)*100 + (minor))
```

Reply:
We need to support projects where people install "deal.II master as of two weeks ago" and "deal.II master as of two months ago" without printing thousands of deprecation warnings so I agree with @kronbichler's general point (its why I closed #10335).

I support doing versioned deprecations. It wouldn't be that hard to have another level of deprecation that can be turned on or off that we promote to `DEAL_II_DEPRECATED` the moment we increment the version number.

Reply:
I would be fine with defining a `DEAL_II_DEPRECATED_EARLY` macro that we promote to `DEAL_II_DEPRECATED` with the next release.

Reply:
Maybe something like https://github.com/dealii/dealii/pull/11472?